src/Fields/sum_of_fields.jl

```diff
@@ -0,0 +1,79 @@
+using Base: @propagate_inbounds
+using Oceananigans.Fields: AbstractField, location
+
+import Adapt: adapt_structure
+import Base: getindex, show, summary
+import Oceananigans.BoundaryConditions: fill_halo_regions!
+import Oceananigans.ImmersedBoundaries: mask_immersed_field!, mask_immersed_field_xy!
+
+"""
+    SumOfFields{N, F}
+
+`SumOfFields` objects hold `N` field/fields and return their sum when indexed.
+"""
+struct SumOfFields{N, LX, LY, LZ, G, T, F} <: AbstractField{LX, LY, LZ, G, T, 3}
+    fields :: F
+      grid :: G
+    
+    function SumOfFields{N}(fields...) where N
+        grid = first(fields).grid
+        loc = location(first(fields))
+
+        all(f.grid == grid for f in fields) || 
+            throw(ArgumentError("All `fields` must be on the same grid"))
+
+        all(location(f) == loc for f in fields) || 
+            throw(ArgumentError("All `fields` must be on the same location"))
+
+        T = eltype(first(fields).data)
+        F = typeof(fields)
+        G = typeof(grid)
+
+        return new{N, loc..., G, T, F}(fields, grid)
+    end
+end
+
+adapt_structure(to, sum::SumOfFields{N}) where N = SumOfFields{N}((adapt_structure(to, field) for field in sum.fields)...)
+
+summary(::SumOfFields{N, LX, LY, LZ}) where {N, LX, LY, LZ}  = "Sum of $N fields at ($LX, $LY, $LZ)"
```

Comment:
This summary does not state the name of the type (`SumOfFields`). The purpose of `summary` and `show` is to give information about the object so I think it's important to state the name.

Also the `summary` should be more uniform with other field summaries

Reply:
I suggest this

```julia
function Base.summary(field::SumOfFields{N}) where N
    LX, LY, LZ = location(field)
    prefix = string(size_summary(size(field)), " SumOfFields{$LX, $LY, $LZ} with $N fields")

    reduced_dims = reduced_dimensions(field)

    suffix = reduced_dims === () ?
        string(" on ", grid_name(field), " on ", summary(architecture(field))) :
        string(" reduced over dims = ", reduced_dims,
               " on ", grid_name(field), " on ", summary(architecture(field)))

    return string(prefix, suffix)
end
```

This is plagiarized from `summary` for `Fields`.

Reply:
There's also a nice `show` method for NamedTuple of Fields that could inspire a `show` method for `SumOfFields`